Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/internal/triangle_soup_snap_rounding.h

```diff
@@ -0,0 +1,358 @@
+*   \cgalParamNBegin{point_map}
+*     \cgalParamDescription{a property map associating points to the elements of the range `soup_points`}
+*     \cgalParamType{a model of `ReadWritePropertyMap` whose value type is a point type}
+*     \cgalParamDefault{`CGAL::Identity_property_map`}
+*   \cgalParamNEnd
+*   \cgalParamNBegin{geom_traits}
+*     \cgalParamDescription{an instance of a geometric traits class}
+*     \cgalParamType{a class model of `Kernel`}
+*     \cgalParamDefault{a \cgal Kernel deduced from the point type, using `CGAL::Kernel_traits`}
+*     \cgalParamExtra{The geometric traits class must be compatible with the point type.}
+*   \cgalParamNEnd
+*   \cgalParamNBegin{snap_grid_size}
+*     \cgalParamDescription{Scale the points to [-2^gs, 2^gs] where gs is the snap_grid_size before to round them on integer.}
+*     \cgalParamType{unsigned int}
+*     \cgalParamDefault{23}
+*     \cgalParamExtra{Must be lower than 52.}
+*   \cgalParamNEnd
+*   \cgalParamNBegin{number_of_iterations}
+*     \cgalParamDescription{Maximum number of iteration performed by the algorithm.}
+*     \cgalParamType{unsigned int}
+*     \cgalParamDefault{15}
+*   \cgalParamNEnd
+* \cgalNamedParamsEnd
+*
+* \return `true` if the modified triangle soup is free from self-intersection, and `false` if the algorithm was not
+* able to provide such a triangle soup within the number of iterations.
+*/
+template <typename PointRange, typename PolygonRange, class NamedParameters>
+bool polygon_soup_snap_rounding(PointRange &points,
+                       PolygonRange &triangles,
+                       const NamedParameters& np)
+{
+  using parameters::choose_parameter;
+  using parameters::get_parameter;
+
+  typedef typename internal_np::Lookup_named_param_def <
+    internal_np::concurrency_tag_t,
+    NamedParameters,
+    Sequential_tag
+  > ::type Concurrency_tag;
+
+
+  constexpr bool parallel_execution = std::is_same_v<Parallel_tag, Concurrency_tag>;
+
+#ifndef CGAL_LINKED_WITH_TBB
+  static_assert (!parallel_execution,
+                 "Parallel_tag is enabled but TBB is unavailable.");
+#endif
+
+  using Point_3 = std::remove_cv_t<typename std::iterator_traits<typename PointRange::const_iterator>::value_type>;
+  using Kernel = typename Kernel_traits<Point_3>::Kernel;
+
+  // Get the grid size from the named parameter, the grid size could not be greater than 52
+  const unsigned int grid_size = (std::min)(52,choose_parameter(get_parameter(np, internal_np::snap_grid_size), 23));
+  const unsigned int max_nb_of_iteration = choose_parameter(get_parameter(np, internal_np::number_of_iterations), 15);
+
+#ifdef PMP_ROUNDING_VERTICES_IN_POLYGON_SOUP_VERBOSE
+  std::cout << "Compute the scaling of the coordinates" << std::endl;
+  std::cout << grid_size << std::endl;
+#endif
+
+  auto exp = [](const double v)
+  {
+    int n;
+    frexp(v, &n);
+    return n;
+  };
+  auto pow_2 = [](const int k)
+  {
+      return (k>=0)?std::pow(2,k):1./std::pow(2,-k);
```

Comment:
I think for performance reasons you better do a shift operation.

Reply:
+ it's done only once for all the vertices.